Start work on on Inky Impression

lawik opened the pull request.

lawik committed:
Start work on on Inky Impression

lawik committed:
More work on setup and update for Inky Impression

lawik committed:
Add all of the Impression code but not working, added cs0 but that won't open

axelson committed:
Kinda working!

axelson committed:
Kind of like a flag

Fixes a off by one error

axelson committed:
Cleaning

axelson committed:
Clean up cs pin calls

Just put a comment instead of an actual call in case we want to restore
the calls later.

axelson committed:
clean

lawik committed:
Merge pull request #42 from axelson/jax/inky-impression

Jax/inky impression

nyaray commented on lib/hal/impression/rpiio.ex:
```diff
@@ -0,0 +1,169 @@
+    {:ok, reset_pid} = gpio.open(pin_mappings[:reset_pin], :output, initial_value: 1)
+    IO.puts("opening busy pin")
+    {:ok, busy_pid} = gpio.open(pin_mappings[:busy_pin], :input)
+    IO.puts("opening SPI device")
```
Do we want to keep the prints in here?

jasonmj replied:
I left them in because it seemed useful and credo wasn't opposed to it either.

nyaray commented on lib/inky.ex:
```diff
@@ -33,6 +34,21 @@ defmodule Inky do
+
+    IO.puts("Seting pixels...")
+    Inky.set_pixels(Inky.Foo, %{})
+  end
```
Looks like a dev function snuck its way in.

jasonmj replied:
Yep, removing this block during my cleanup.

nyaray commented on lib/hal/hal.ex:
```diff
@@ -12,5 +12,5 @@ defmodule Inky.HAL do
               state :: Inky.IOCommands.State.t()
             ) ::
-              :ok | {:error, :device_busy}
+              io_state() | :ok | {:error, :device_busy}
```
Not cool, hehe... why leak `any()` from here?

jasonmj replied:
Good call, reverting in cleanup.

nyaray commented on lib/hal/impression/rpihal.ex:
```diff
@@ -0,0 +1,319 @@
+    :yellow => 5,
+    :orange => 6,
+    # Not a color, but is used to clear the display
+    :clean => 7
```
Should probably be named `clear`

nyaray commented on lib/display/display.ex:
```diff
@@ -4,18 +4,31 @@ defmodule Inky.Display do
+
   @type t() :: %__MODULE__{}
 
-  @enforce_keys [:type, :width, :height, :packed_dimensions, :rotation, :accent, :luts]
```
If these are required for one display but not the other, we should consider refactoring things so that we don't force re-usability where it is not required/meaningful (maybe that's exactly what you did here? :D). "False generalisation" only makes things unclear since you don't knew when/if something is required/missing.

jasonmj replied:
I've reverted this change too, since these keys can peacefully coexist in the struct the defines the impression display. We're not using packed dimensions because it's redundant/unnecessary, but if someone wants to use them in the future, it'll be possible.

As for luts, it's not used at all for the impression. Not sure we need to define a separate struct+key enforcement just for that though.

nyaray replied:
This is good food for thought, though... if the devices are diverting, perhaps our configurations should, too.

One way of dealing with it would be to have a protocol common for all types, for the things that are common to all displays and then have the specific drivers check if it is a value of the correct underlying type at runtime during init. But that might be a bit more than what you signed up for at this point and a bit of over-engineering before we have several displays where the fit is off... thoughts?

jasonmj replied:
That sounds like it might be a worthwhile refactor to open a separate issue for.

nyaray commented on lib/display/display.ex:
```diff
@@ -83,4 +96,32 @@ defmodule Inky.Display do
+      [177, 106, 73],
+      [255, 255, 255]
+    ]
+  end
```
Feels like things are getting  muddled in display.ex... we might want to consider how things could be split up. Not sure if it should be by amount of colours or "base type" of display. The what/phat share things, but I'm seeing quite a few things that aren't really compatible with the impression.

nyaray commented on lib/hal/impression/rpihal.ex:
```diff
@@ -0,0 +1,319 @@
+          true -> <<rem(y, @colors[:orange] + 1)>>
+        end
+        # color = floor(0 / w * 7)
+      end
```
what. is this a dev-demo function?

nyaray commented on lib/hal/impression/rpihal.ex:
```diff
@@ -0,0 +1,319 @@
+      end
+
+    log("Generated buffer of size: #{byte_size(buffer)}")
+    log("buffer: #{inspect(buffer)}")
```
might not want to do this...

nyaray commented on lib/hal/impression/rpihal.ex:
```diff
@@ -0,0 +1,319 @@
+        0x23,
+        # UC8159_7C
+        0x23
+      ])
```
Hahaha, omg, what is this :D. This should probably be cleaned up with parens, `use Bitwise` and the `|||` operator.

nyaray replied:
Wait a minute... if the shifting/bor hackery is done to build a complex binary by glueing together values, we can do that in a much simpler binary expression.

jasonmj replied:
I went for pipes instead of `|||`. Seems to accommodate the comments nicely.

nyaray commented on lib/display/display.ex:
```diff
@@ -4,18 +4,31 @@ defmodule Inky.Display do
+      type: type,
+      width: 600,
+      height: 448,
+      packed_resolution: <<2, 88, 1, 192>>, # I used the struct.pack in Python to generate this
```
Some observations...

`(2*255 + 88) = 600-2 = 598`
`(1*255 + 192) = 488-1 = 447`

How do these numbers relate to whatever input you put into python? Also, if we're not going to do this properly in elixir, we should at the very least have the actual code here, not just a mention of "it was done in python".

The dimensions for the Inky were kept in a struct, why not do the same thing here and convert it to a binary when it's required?

nyaray commented on lib/display/display.ex:
```diff
@@ -83,4 +96,32 @@ defmodule Inky.Display do
+      [177, 106, 73],
+      [255, 255, 255]
+    ]
+  end
```
What exactly is the purpose of these two colorsets? Are they operating modes of the display? :O

jasonmj replied:
Looks like they come from right here: https://github.com/pimoroni/inky/blob/54684464b2f35bfd52208cdfb922c09685644181/library/inky/inky_uc8159.py#L26

nyaray replied:
Do the colours actually get sent to the monitor at some point? @lawik

lawik replied:
I don't remember if I added them or Jason did :D

jasonmj replied:
Yeah... these don't get used anywhere. I'm reverting during my cleanup.

nyaray commented on lib/hal/impression/rpihal.ex:
```diff
@@ -0,0 +1,319 @@
+        # UC8159_7C
+        0x23,
+        # UC8159_7C
+        0x23
```
twice?

jasonmj replied:
Good catch! This issue is also present in the python driver, but it works fine without a second command.

nyaray commented on lib/hal/impression/rpihal.ex:
```diff
@@ -0,0 +1,319 @@
+  defp write_command(state, command, data) do
+    io_call(state, :handle_command, [command, data])
+    state
+  end
```
This is why we don't copy-paste stuff 😅

jasonmj replied:
These are all being used. Is there an outstanding issue here?

nyaray commented on lib/inky.ex:
```diff
@@ -173,7 +206,16 @@ defmodule Inky do
+        :ok -> state
+        {:error, _} -> state
+        {:ok, state} -> state
+      end
```
Dropping errors silently, are we now? :O

jasonmj replied:
Yeah, apparently this change is no longer needed... reverting.

nyaray commented on lib/inky.ex:
```diff
@@ -185,7 +227,13 @@ defmodule Inky do
+        {:error, _} -> state
+        {:ok, state} -> state
+      end
+
```
More silent failures?

jasonmj replied:
This change is no longer needed either... reverting.

jasonmj commented on lib/hal/impression/rpiio.ex:
```diff
@@ -87,7 +49,7 @@ defmodule Inky.Impression.RpiIO do
   }
 
-  @spi_speed_hz 488_000
+  @spi_speed_hz 3000000
```
Maybe 3_000_000 instead?

jasonmj reviewed (commented):
Got a quick look at the work on this branch. Not a lot to add here, but excited to see progress. Looking forward to getting my impression installed on a pi soon to join in on the fun :)

Commit:
Fix log typo

Commit:
Adds comments to byte commands defined for Impression in hal

Commit:
Fix typo in comment

Commit:
Update mix deps for development on jasonmj

Commit:
Update display definition for impression

Commit:
Add handling for impression in packed_width and packed_height

Commit:
Fix usage of state in log function

Commit:
Add set_dimensions to do_update for impression

Commit:
Replace buffer generation with PixelUtil.pixels_to_bits

Commit:
Add a color_map for the impression

Commit:
Add notes about WIP toward impression support

Commit:
Reverting dev changes to mix.exs

Commit:
Cleaning up unused code

Commit:
Adding bit_size option to PixelUtil.pixels_to_bits

lawik committed:
Merge pull request #46 from jasonmj/impression

Impression

Commit:
Cleans up code for PR

Commit:
Adds painter test

jasonmj commented:
@lawik @nyaray I've worked to clean up this PR and responded on all of the comments. Please get a look and see if there's more you'd like done here before approving to merge. Thanks!

axelson committed:
Support circuits_gpio 1.0

There's no code changes, only the version number has been updated:
https://github.com/elixir-circuits/circuits_gpio/blob/cc7d77479a4d2ffba1a24962ad067f4680003de4/CHANGELOG.md#v100---10-20-2021

Commit:
Merge pull request #47 from axelson/impression

Support circuits_gpio 1.0

Commit:
Add support for buttons on Inky Impression

Commit:
Support sending button messages to a process by registered atom name

lawik committed:
Merge pull request #48 from jasonmj/impression

Add support for buttons on Inky Impression

axelson commented:
This can probably be closed in favor of #49
